File: nodejs/src/multicall.ts
import type { PublicClient, Transport } from "viem"
import type { base } from "viem/chains"
import { AFI_ADDRESS, ERC20_ABI, MULTICALL3_ADDRESS } from "./constants.js"
import type { Address, TokenInfo } from "./types.js"

type BasePublicClient = PublicClient<Transport, typeof base>

const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000" as Address

function emptyOwner(owner: Address | undefined): boolean {
  return !owner || owner.toLowerCase() === ZERO_ADDRESS
}

/** Immutable per-token metadata that the cache can safely reuse forever. */
export interface TokenMetadata {
  symbol:   string
  name:     string
  decimals: number
}

/** Public type for a single contract call passed to `client.multicall()`. */
export interface MulticallContract {
  address:      Address
  abi:          readonly unknown[]
  functionName: string
  args?:        readonly unknown[]
}

/** Public type for a single multicall result. */
export type MulticallResult =
  | { status: "success"; result: unknown }
  | { status: "failure"; error: Error }

/**
 * Generic multicall wrapper. Bundles N read calls into a single RPC round-trip
 * via Multicall3. Use for arbitrary contract reads beyond what the SDK exposes.
 */
export async function genericMulticall(
  client: BasePublicClient,
  contracts: MulticallContract[],
  opts: { allowFailure?: boolean } = {},
): Promise<MulticallResult[]> {
  const allowFailure = opts.allowFailure ?? true
  const results = (await client.multicall({
    contracts: contracts as any,
    multicallAddress: MULTICALL3_ADDRESS,
    allowFailure,
  })) as Array<{ status: "success"; result: unknown } | { status: "failure"; error?: Error }>
  return results.map((r) =>
    r.status === "success"
      ? { status: "success", result: r.result }
      : { status: "failure", error: r.error ?? new Error("call reverted") },
  )
}

/**
 * Fetches token metadata in a single multicall.
 * When `owner` is provided, also includes balance + allowance against the AFI contract.
 *
 * `metaCache` (optional) caches immutable metadata across calls — pass the client's
 * shared cache to avoid re-fetching symbol/name/decimals for repeated lookups.
 */
export async function fetchTokenInfo(
  token: Address,
  client: BasePublicClient,
  owner?: Address,
  metaCache?: Map<string, TokenMetadata>,
): Promise<TokenInfo> {
  const [info] = await fetchTokenInfoBatch([token], client, owner, metaCache)
  return info
}

/**
 * Batched form of fetchTokenInfo. Returns results in the same order as the input.
 * When `metaCache` is provided, tokens with cached metadata only fetch
 * balance/allowance (or nothing, if no owner is given).
 */
export async function fetchTokenInfoBatch(
  tokens: Address[],
  client: BasePublicClient,
  owner?: Address,
  metaCache?: Map<string, TokenMetadata>,
): Promise<TokenInfo[]> {
  if (tokens.length === 0) return []
  const withOwner = !emptyOwner(owner)

  // Partition tokens by whether their metadata is already cached.
  type Plan = { token: Address; cached: TokenMetadata | undefined; offset: number; count: number }
  const plans: Plan[] = []
  const contracts: MulticallContract[] = []
  for (const token of tokens) {
    const cached = metaCache?.get(token.toLowerCase())
    const plan: Plan = { token, cached, offset: contracts.length, count: 0 }
    if (!cached) {
      contracts.push(
        { address: token, abi: ERC20_ABI, functionName: "symbol" },
        { address: token, abi: ERC20_ABI, functionName: "name" },
        { address: token, abi: ERC20_ABI, functionName: "decimals" },
      )
      plan.count += 3
    }
    if (withOwner) {
      contracts.push(
        { address: token, abi: ERC20_ABI, functionName: "balanceOf", args: [owner!] },
        { address: token, abi: ERC20_ABI, functionName: "allowance", args: [owner!, AFI_ADDRESS] },
      )
      plan.count += 2
    }
    plans.push(plan)
  }

  // If nothing to fetch (everything cached and no owner), return purely from cache.
  const results: MulticallResult[] =
    contracts.length > 0
      ? await genericMulticall(client, contracts, { allowFailure: true })
      : []

  return plans.map((plan) => {
    let cursor = plan.offset
    let meta: TokenMetadata
    if (plan.cached) {
      meta = plan.cached
    } else {
      const symRes = results[cursor++]
      const nameRes = results[cursor++]
      const decRes = results[cursor++]
      if (decRes.status !== "success") throw new Error(`decimals() reverted for ${plan.token}`)
      meta = {
        symbol:   symRes.status === "success" ? (symRes.result as string) : "",
        name:     nameRes.status === "success" ? (nameRes.result as string) : "",
        decimals: Number(decRes.result as number | bigint),
      }
      metaCache?.set(plan.token.toLowerCase(), meta)
    }

    const info: TokenInfo = { address: plan.token, ...meta }
    if (withOwner) {
      const balRes = results[cursor++]
      const allowRes = results[cursor++]
      if (balRes.status !== "success") throw new Error(`balanceOf() reverted for ${plan.token}`)
      if (allowRes.status !== "success") throw new Error(`allowance() reverted for ${plan.token}`)
      info.owner = owner
      info.balance = balRes.result as unknown as bigint
      info.allowance = allowRes.result as unknown as bigint
    }
    return info
  })
}

File: go/multicall.go
package afi

import (
	"context"
	"fmt"
	"sync"

	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"
)

// Multicall3Call is a single call submitted to Multicall3.aggregate3.
type Multicall3Call struct {
	Target       common.Address `json:"target"`
	AllowFailure bool           `json:"allowFailure"`
	CallData     []byte         `json:"callData"`
}

// Multicall3Result is one entry returned by Multicall3.aggregate3.
type Multicall3Result struct {
	Success    bool   `json:"success"`
	ReturnData []byte `json:"returnData"`
}

// TokenMetadata is the immutable per-token data that the metadata cache holds.
type TokenMetadata struct {
	Symbol   string
	Name     string
	Decimals uint8
}

// metadataCache is a thread-safe address-keyed cache for TokenMetadata.
type metadataCache struct {
	mu sync.RWMutex
	m  map[common.Address]TokenMetadata
}

func newMetadataCache() *metadataCache {
	return &metadataCache{m: make(map[common.Address]TokenMetadata)}
}

func (c *metadataCache) get(token common.Address) (TokenMetadata, bool) {
	if c == nil {
		return TokenMetadata{}, false
	}
	c.mu.RLock()
	defer c.mu.RUnlock()
	v, ok := c.m[token]
	return v, ok
}

func (c *metadataCache) set(token common.Address, meta TokenMetadata) {
	if c == nil {
		return
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	c.m[token] = meta
}

func (c *metadataCache) clear() {
	if c == nil {
		return
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	c.m = make(map[common.Address]TokenMetadata)
}

// aggregate3 calls Multicall3.aggregate3 with the given calls.
// Each entry's AllowFailure controls whether an individual revert tanks the batch.
func aggregate3(
	ctx context.Context,
	client *ethclient.Client,
	mcABI abi.ABI,
	calls []Multicall3Call,
) ([]Multicall3Result, error) {
	if len(calls) == 0 {
		return []Multicall3Result{}, nil
	}
	input, err := mcABI.Pack("aggregate3", calls)
	if err != nil {
		return nil, fmt.Errorf("pack aggregate3: %w", err)
	}

	raw, err := client.CallContract(ctx, ethereum.CallMsg{
		To:   &Multicall3Address,
		Data: input,
	}, nil)
	if err != nil {
		return nil, fmt.Errorf("multicall aggregate3: %w", err)
	}

	var results []Multicall3Result
	if err := mcABI.UnpackIntoInterface(&results, "aggregate3", raw); err != nil {
		return nil, fmt.Errorf("unpack aggregate3: %w", err)
	}
	return results, nil
}
